src/flow/assembly_lmh.hh

```diff
@@ -235,139 +232,144 @@ protected:
-        
-        Boundary *bcd;
-        unsigned int side_row, edge_row;
         
-        dirichlet_edge.resize(nsides);
-        for (unsigned int i = 0; i < nsides; i++) {
-
-            side_row = loc_side_dofs[i];    //local
-            edge_row = loc_edge_dofs[i];    //local
-            
-            bcd = ele.side(i)->cond();
-            dirichlet_edge[i] = 0;
-            if (bcd) {
-                ElementAccessor<3> b_ele = bcd->element_accessor();
-                DarcyMH::EqData::BC_Type type = (DarcyMH::EqData::BC_Type)ad_->bc_type.value(b_ele.centre(), b_ele);
+        dirichlet_edge.resize(ele->n_sides());
+        for(DHCellSide side : dh_cell.side_range()){
+            unsigned int sidx = side.side_idx();
+            dirichlet_edge[sidx] = 0;
 
+            // assemble BC
+            if (side.cond()) {
                 double cross_section = ad_->cross_section.value(ele.centre(), ele);
+                assemble_side_bc(side, cross_section, use_dirichlet_switch);
 
-                ad_->balance->add_flux_matrix_values(ad_->water_balance_idx, ele.side(i),
-                                                    {global_dofs_[loc_side_dofs[i]]}, {1});
+                ad_->balance->add_flux_values(ad_->water_balance_idx, side,
+                                              {loc_system_.row_dofs[loc_side_dofs[sidx]]},
+                                              {1}, 0);
+            }
 
-                if ( type == DarcyMH::EqData::none) {
-                    // homogeneous neumann
-                } else if ( type == DarcyMH::EqData::dirichlet ) {
-                    double bc_pressure = ad_->bc_pressure.value(b_ele.centre(), b_ele);
-                    loc_schur_.set_solution(i, bc_pressure);
-                    dirichlet_edge[i] = 1;
-                    
-                } else if ( type == DarcyMH::EqData::total_flux) {
-                    // internally we work with outward flux
-                    double bc_flux = -ad_->bc_flux.value(b_ele.centre(), b_ele);
-                    double bc_pressure = ad_->bc_pressure.value(b_ele.centre(), b_ele);
-                    double bc_sigma = ad_->bc_robin_sigma.value(b_ele.centre(), b_ele);
-                    
-                    dirichlet_edge[i] = 2;  // to be skipped in LMH source assembly
-                    loc_system_.add_value(edge_row, edge_row,
-                                            -b_ele.measure() * bc_sigma * cross_section,
-                                            (bc_flux - bc_sigma * bc_pressure) * b_ele.measure() * cross_section);
+            // side-edge (flux-lambda) terms
+            loc_system_.add_value(loc_side_dofs[sidx], loc_edge_dofs[sidx], 1.0);
+            loc_system_.add_value(loc_edge_dofs[sidx], loc_side_dofs[sidx], 1.0);
+        }
+    }
+
+
+    void assemble_side_bc(const DHCellSide& side, double cross_section, bool use_dirichlet_switch)
+    {
+        const unsigned int sidx = side.side_idx();
```

Comment:
I hope this doesn't change as the diff is a mess.

Reply:
I did this with intention not to do any changes in the code logic, but to separate the switch part of the BC code and to reduce the tabs of all the "embeddings".